core/shared_queue.rs

```diff
@@ -0,0 +1,229 @@
+// Copyright 2018 the Deno authors. All rights reserved. MIT license.
+use crate::isolate::Buf;
+use crate::libdeno::deno_buf;
+
+const MAX_RECORDS: usize = 100;
+/// Total number of records added.
+const INDEX_NUM_RECORDS: usize = 0;
+/// Number of records that have been shifted off.
+const INDEX_NUM_SHIFTED_OFF: usize = 1;
+/// The head is the number of initialized bytes in SharedQueue.
+/// It grows monotonically.
+const INDEX_HEAD: usize = 2;
+const INDEX_OFFSETS: usize = 3;
+const INDEX_RECORDS: usize = 3 + MAX_RECORDS;
+/// Byte offset of where the records begin. Also where the head starts.
+const HEAD_INIT: usize = 4 * INDEX_RECORDS;
+/// A rough guess at how big we should make the shared buffer in bytes.
+pub const RECOMMENDED_SIZE: usize = 128 * MAX_RECORDS;
+
+pub struct SharedQueue {
+  bytes: Vec<u8>,
+}
+
+impl SharedQueue {
+  pub fn new(len: usize) -> Self {
+    let mut bytes = Vec::new();
+    bytes.resize(HEAD_INIT + len, 0);
+    let mut q = Self { bytes };
+    q.reset();
+    q
+  }
+
+  pub fn as_deno_buf(&self) -> deno_buf {
+    let ptr = self.bytes.as_ptr();
+    let len = self.bytes.len();
+    unsafe { deno_buf::from_raw_parts(ptr, len) }
+  }
+
+  /// Clears the shared buffer.
+  pub fn reset(&mut self) {
+    let s: &mut [u32] = self.as_u32_slice_mut();
+    for i in 0..INDEX_RECORDS {
+      s[i] = 0;
+    }
+    s[INDEX_HEAD] = HEAD_INIT as u32;
+  }
+
+  fn as_u32_slice<'a>(&'a self) -> &'a [u32] {
+    let p = self.bytes.as_ptr() as *const u32;
+    unsafe { std::slice::from_raw_parts(p, self.bytes.len() / 4) }
+  }
+
+  fn as_u32_slice_mut<'a>(&'a mut self) -> &'a mut [u32] {
+    let p = self.bytes.as_mut_ptr() as *mut u32;
+    unsafe { std::slice::from_raw_parts_mut(p, self.bytes.len() / 4) }
+  }
+
+  pub fn size(&self) -> usize {
+    let s = self.as_u32_slice();
+    (s[INDEX_NUM_RECORDS] - s[INDEX_NUM_SHIFTED_OFF]) as usize
+  }
+
+  fn num_records(&self) -> usize {
+    let s = self.as_u32_slice();
+    s[INDEX_NUM_RECORDS] as usize
+  }
+
+  fn head(&self) -> usize {
+    let s = self.as_u32_slice();
+    s[INDEX_HEAD] as usize
+  }
+
+  fn set_end(&mut self, index: usize, end: usize) {
+    let s = self.as_u32_slice_mut();
+    s[INDEX_OFFSETS + index] = end as u32;
+  }
+
+  fn get_end(&self, index: usize) -> Option<usize> {
+    if index < self.num_records() {
+      let s = self.as_u32_slice();
+      Some(s[INDEX_OFFSETS + index] as usize)
+    } else {
+      None
+    }
+  }
+
+  fn get_offset(&self, index: usize) -> Option<usize> {
+    if index < self.num_records() {
+      Some(if index == 0 {
+        HEAD_INIT
+      } else {
+        let s = self.as_u32_slice();
+        s[INDEX_OFFSETS + index - 1] as usize
+      })
+    } else {
+      None
+    }
+  }
+
+  /// Returns none if empty.
+  pub fn shift<'a>(&'a mut self) -> Option<Buf> {
```

Comment:
I think it'd be better to return an Option<&mut [u8]>. This avoids allocation and copy.

Reply:
But what should the lifetime be? just set it to `'a`? it's not quite correct - but maybe correct enough for us...

Reply:
You can do `pub fn shift<'a: 'b, 'b>(&'a mut self) -> Option<&'b mut [u8]>`

Reply:
In other words, there are two lifetimes. The returned borrow's lifetime must be no longer than `self`'s lifetime.

Reply:
I just pushed a change to `pub fn shift<'a>(&'a mut self) -> Option<&'a [u8]>` 

It does seem to be a hair faster with this.